src/pybind/mgr/balancer/module.py

```diff
@@ -33,63 +274,487 @@ def shutdown(self):
-            mode = self.get_config('mode', default_mode)
+            self.log.debug('Waking up')
+            self.active = self.get_config('active', '') is not ''
             sleep_interval = float(self.get_config('sleep_interval',
                                                    default_sleep_interval))
-            max_misplaced = float(self.get_config('max_misplaced',
-                                                  default_max_misplaced))
-            self.log.info('Mode %s, sleep interval %f, max misplaced %f' %
-                          (mode, sleep_interval, max_misplaced))
-
-            info = self.get('pg_status')
-            if info.get('unknown_ratio', 0.0) + info.get('degraded_ratio', 0.0) + info.get('inactive_ratio', 0.0) > 0.0:
-                self.log.info('Some PGs are unknown, degraded, or inactive; waiting')
-            elif info.get('misplaced_ratio', 0.0) >= max_misplaced:
-                self.log.info('Too many PGs (%f > max %f) are misplaced; waiting' % (info.get('misplaced_ratio', 0.0), max_misplaced))
-            else:
-                if mode == 'upmap':
-                    self.do_upmap()
-                elif mode == 'crush':
-                    self.do_crush(compat=False)
-                elif mode == 'crush-compat':
-                    self.do_crush(compat=True)
-                elif mode == 'osd_weight':
-                    self.osd_weight()
-                elif mode == 'none':
-                    self.log.info('Idle')
-                else:
-                    self.log.info('Unrecognized mode %s' % mode)
+            if self.active:
+                self.log.debug('Running')
+                plan = self.plan_create('auto-foo')
+                self.optimize(plan)
+                #self.plan_apply(plan)
+            self.log.debug('Sleeping for %d', sleep_interval)
             self.event.wait(sleep_interval)
+            self.event.clear()
+
+    def plan_create(self, name):
+        plan = Plan(name, MappingState(self.get_osdmap(),
+                                       self.get("pg_dump"),
+                                       'plan %s initial' % name))
+        self.plans[name] = plan
+        return plan
+
+    def plan_rm(self, name):
+        if name in self.plans:
+            del self.plans[name]
+
+    def calc_eval(self, ms):
+        pe = Eval(ms)
+        pool_rule = {}
+        for p in ms.osdmap_dump.get('pools',[]):
+            pe.pool_name[p['pool']] = p['pool_name']
+            pe.pool_id[p['pool_name']] = p['pool']
+            pool_rule[p['pool_name']] = p['crush_rule']
+            pe.pool_roots[p['pool_name']] = []
+        pools = pe.pool_id.keys()
+        if len(pools) == 0:
+            return pe
+        self.log.debug('pool_name %s' % pe.pool_name)
+        self.log.debug('pool_id %s' % pe.pool_id)
+        self.log.debug('pools %s' % pools)
+        self.log.debug('pool_rule %s' % pool_rule)
+        
+        # get expected distributions by root
```

Comment:
this should support multiple roots per rule by looping on the rules instead of the roots maybe ?